gcc/rust/hir/rust-ast-lower-type.cc

```diff
@@ -17,14 +17,42 @@
 // <http://www.gnu.org/licenses/>.
 
 #include "rust-ast-lower-type.h"
-#include "optional.h"
-#include "rust-attribute-values.h"
+#include "rust-hir-map.h"
+#include "rust-hir-path.h"
+#include "rust-path.h"
+#include "rust-pattern.h"
 
 namespace Rust {
 namespace HIR {
 
 HIR::TypePath *
-ASTLowerTypePath::translate (AST::TypePath &type)
+ASTLowerTypePath::translate (AST::Path &type)
+{
+  switch (type.get_path_kind ())
+    {
+    case AST::Path::Kind::LangItem:
+      return translate_lang_item_type_path (
+	static_cast<AST::LangItemPath &> (type));
+    case AST::Path::Kind::Type:
+      return translate_type_path (static_cast<AST::TypePath &> (type));
+    case AST::Path::Kind::Regular:
+      rust_unreachable ();
+    }
+}
+
+// FIXME: Before merging: Can we simplify these two with a simple call to visit
+// to do dynamic dispatch?
+HIR::TypePath *
+ASTLowerTypePath::translate_type_path (AST::TypePath &type)
+{
+  ASTLowerTypePath resolver;
+  type.accept_vis (resolver);
+  rust_assert (resolver.translated != nullptr);
+  return resolver.translated;
+}
+
+HIR::TypePath *
+ASTLowerTypePath::translate_lang_item_type_path (AST::LangItemPath &type)
 {
```

Comment:
I feel like this manual dispatch isn't necessary, and that we can just use dynamic dispatch here. I'll try and fix it